tests/snuba/api/endpoints/test_organization_events_stats.py

```diff
@@ -136,26 +136,28 @@ def test_generic_issue_calculated_interval(self):
         """Test that a 4h interval returns the correct generic event stats.
         This follows a different code path than 1h or 1d as the IssuePlatformTimeSeriesQueryBuilder
         does some calculation to create the time column."""
+        # Use an event window that is unaffected by other tests.
```

Comment:
What do you mean by affected by other tests? Snuba should be flushed between each test run, afaik

Reply:
This might be due to caching on the snuba api side. I'd have thought the project ids would change between runs though, which would bust the caching. Shouldn't be anything else that would cause this. I think we'd be better off understanding the root cause of this rather than working around it though.

If you project out the project ids used in these tests are they the same?

Reply:
I did log the qualified_short_id in case there was some re-use there but they are different.

The only value I saw that was the same was the groups `short_id`, but that seems to be mixed with the group id which is unique.

Reply:
What would be most interesting here is to understand if the `project_id` changes between the two tests. If it doesn't, then we'll run the same query in a short period of time, and we might hit the snuba cache.

Reply:
Will take a look when I can

Reply:
Just gave it another run with some more logging, I managed to catch a failing example on my first run and the project ID was the same. For completeness:

First test (test_generic_issue) had the following values:

project id:							`4551382353117185`
user id:								`41`
group id:								`32`
group short id:					`3`
group qualified short id:		`MOVING-PUG-3`	

The second test (test_generic_issue_calculated_interval) has the following values:

project id:							`4551382353117185`
user id:								`44`
group id:								`36`
group short id:					`3`
group qualified short id:		`PERFECT-CAIMAN-3`

Reply:
That's interesting, so it looks like both the `group_short_id` and `project_id` are the same. The query we're making is on the short id https://github.com/getsentry/sentry/blob/eb5ee72fd47a79d2866044f590b249e510ea5b0c/tests/snuba/api/endpoints/test_organization_events_stats.py#L170.

So that'd imply that queries made in the same short period of time would be exactly the same, and would hit the caching problems from snuba. So we really have two options for fixing this more generically:
 - Make sure the project ids aren't the same. They shouldn't be, but I think something to do with snowflake ids can cause them to be the same.
 - Have snuba flush their cache when we reset it

@markstory I remember investigating the snowflake ids a few months back and seeing that there was some bug that caused them to not always change. I think we had a ticket for it, did anything ever happen there?

Reply:
Did a little searching out of curiosity. It looks like these IDs are generated by [snowflake.py](https://github.com/getsentry/sentry/blob/master/src/sentry/utils/snowflake.py) which seems to be a combination of version, region and timestamp.

The timestamp is posix time in seconds, which probably explains why some of the IDs are duplicating.

Running the tests with a global of the previous tests project ID and comparing with the current project ID it's easy to find tests sharing the same ID (20%+ of tests in `test_organization_events_states.py` seem to be using the same ID as the previous test).

Running the test suite with `SENTRY_USE_SNOWFLAKE = False` fixes the issue, presumably because that falls back to a django sequence that doesn't get reset between tests.

Reply:
Yeah, the sequence would definitely be reset between runs since we just flush redis.

So re-reading the thread I suddenly realised that the caching theory didn't make sense, since the short id is part of the query. This translates into a group id, which *is* unique between runs and should produce different queries. I dug into this further, and found that there's a bug where the short id isn't being included in the query. If I add an event associated with different group to the test, then it fails as well since it counts that event even though the query should filter it out.

So taking that into account, then the flakiness is likely still caching, since the group id isn't involved here. I'll fix this specific bug, which will by also fix this flakiness, since we now have a unique value that changes between runs. But this still leaves the real root cause lurking to bite us again in a few months, so it'd be nice to either have a way to flush the cache in snuba, or have our snowflake ids always be unique in between test runs.

Reply:
I'm glad you figured this out.

I tried running the test the other day without the snowflakes and it actually made the flakey behavior easier to trigger, I assumed I'd done something wrong, but it sounds like it's aligning with your findings.

I wonder if there is a way to include part of the pytest environment with the snowflake ID if it's defined and use that as an additional key - quick search shows this: https://docs.pytest.org/en/latest/example/simple.html#pytest-current-test-environment-variable

Reply:
Ok, so I realised that we hadn't added our new dataset to the snuba flushing code. I ran into a flaky test while working on https://github.com/getsentry/sentry/pull/44442 and fixed it there.

So I imagine that is the root cause here. I think that the shared project ids probably exacerbated the issues, since we probably wouldn't have hit it if we'd have separate project ids.

Reply:
I tried running the test with `SENTRY_USE_SNOWFLAKE=FALSE pytest tests/snuba/api/endpoints/test_organization_events_stats.py` to ensure unique project IDs and the test reliably fails, so I think the project ID may be a red herring.